File: Server/Models/Game/Players/Player.cs
using Server.Models.Game.Timeline;
using System.Text.Json.Serialization;

namespace Server.Models.Game.Players; 
[JsonDerivedType(typeof(Player), "Standard")]
[JsonDerivedType(typeof(OnlinePlayer), "Online")]
public class Player
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = string.Empty;
    public int Tokens { get; set; } = 2; // start with 2 tokens
    public List<TimelineCard> Timeline { get; set; } = new(); 
    public bool HasWon => Timeline.Count >= 10; 
    public Guid BaseGameSessionId { get; set; }
    
}

File: Client/src/types/player.ts
import type { TimelineCard } from './song'; 

// Standard Player Type
export interface Player {
  id: string; // Guid -> string
  name: string;
  tokens: number;
  timeline: TimelineCard[]; // List -> Array
  hasWon: boolean; 
  baseGameSessionId: string;
  joinOrder: number;
} 

// Online Player Type (extending Standard Player)
export interface OnlinePlayer extends Player {
  connectionId: string | null; // string? -> string | null
  isHost: boolean;
  isConnected: boolean;
}